fish-rust/src/builtins/tests/string_tests.rs

```diff
@@ -0,0 +1,285 @@
+    string_test!(["string", "sub", "-s6", "abcde"], STATUS_CMD_OK, "\n");
+    string_test!(["string", "sub", "-s-1", "abcde"], STATUS_CMD_OK, "e\n");
+    string_test!(["string", "sub", "-s-5", "abcde"], STATUS_CMD_OK, "abcde\n");
+    string_test!(["string", "sub", "-s-6", "abcde"], STATUS_CMD_OK, "abcde\n");
+    string_test!(["string", "sub", "-s1", "-l0", "abcde"], STATUS_CMD_OK, "\n");
+    string_test!(["string", "sub", "-s1", "-l1", "abcde"], STATUS_CMD_OK, "a\n");
+    string_test!(["string", "sub", "-s2", "-l2", "abcde"], STATUS_CMD_OK, "bc\n");
+    string_test!(["string", "sub", "-s-1", "-l1", "abcde"], STATUS_CMD_OK, "e\n");
+    string_test!(["string", "sub", "-s-1", "-l2", "abcde"], STATUS_CMD_OK, "e\n");
+    string_test!(["string", "sub", "-s-3", "-l2", "abcde"], STATUS_CMD_OK, "cd\n");
+    string_test!(["string", "sub", "-s-3", "-l4", "abcde"], STATUS_CMD_OK, "cde\n");
+    string_test!(["string", "sub", "-q"], STATUS_CMD_ERROR, "");
+    string_test!(["string", "sub", "-q", "abcde"], STATUS_CMD_OK, "");
+
+    string_test!(["string", "trim"], STATUS_CMD_ERROR, "");
+    string_test!(["string", "trim", ""], STATUS_CMD_ERROR, "\n");
+    string_test!(["string", "trim", " "], STATUS_CMD_OK, "\n");
+    string_test!(["string", "trim", "  \x0C\n\r\t"], STATUS_CMD_OK, "\n");
+    string_test!(["string", "trim", " a"], STATUS_CMD_OK, "a\n");
+    string_test!(["string", "trim", "a "], STATUS_CMD_OK, "a\n");
+    string_test!(["string", "trim", " a "], STATUS_CMD_OK, "a\n");
+    string_test!(["string", "trim", "-l", " a"], STATUS_CMD_OK, "a\n");
+    string_test!(["string", "trim", "-l", "a "], STATUS_CMD_ERROR, "a \n");
+    string_test!(["string", "trim", "-l", " a "], STATUS_CMD_OK, "a \n");
+    string_test!(["string", "trim", "-r", " a"], STATUS_CMD_ERROR, " a\n");
+    string_test!(["string", "trim", "-r", "a "], STATUS_CMD_OK, "a\n");
+    string_test!(["string", "trim", "-r", " a "], STATUS_CMD_OK, " a\n");
+    string_test!(["string", "trim", "-c", ".", " a"], STATUS_CMD_ERROR, " a\n");
+    string_test!(["string", "trim", "-c", ".", "a "], STATUS_CMD_ERROR, "a \n");
+    string_test!(["string", "trim", "-c", ".", " a "], STATUS_CMD_ERROR, " a \n");
+    string_test!(["string", "trim", "-c", ".", ".a"], STATUS_CMD_OK, "a\n");
+    string_test!(["string", "trim", "-c", ".", "a."], STATUS_CMD_OK, "a\n");
+    string_test!(["string", "trim", "-c", ".", ".a."], STATUS_CMD_OK, "a\n");
+    string_test!(["string", "trim", "-c", "\\/", "/a\\"], STATUS_CMD_OK, "a\n");
+    string_test!(["string", "trim", "-c", "\\/", "a/"], STATUS_CMD_OK, "a\n");
+    string_test!(["string", "trim", "-c", "\\/", "\\a/"], STATUS_CMD_OK, "a\n");
+    string_test!(["string", "trim", "-c", "", ".a."], STATUS_CMD_ERROR, ".a.\n");
+
+    let saved_flag = feature_test(FeatureFlag::qmark_noglob);
+    unsafe { mutable_fish_features().as_mut() }.unwrap().set(FeatureFlag::qmark_noglob, true);
```

Comment:
This will be a potential problem due to race-conditions in tests. I suggest wrapping the static `Feature` in a https://docs.rs/parking_lot/latest/parking_lot/type.RwLock.html, so that tests like these can turn into:
```rust
{
    // start write scope
    let features = FEATURES.write();
    let saved_flag = features.test(FeatureFlag::mark_noglob);
    let features = features.downgrade_to_upgradeable();
    // do test
    let features = features.upgrade();
    features.set(FeatureFlag::mark_noglob, saved_flag);
}
```